### DISCUSSION TITLE
How do I debug a grammar?

### DISCUSSION THREAD

**Discussion by verhovsky** [tree-sitter/tree-sitter#2218] [Q&A] [answered]
I want to know what course the parser took through a given input and why it applied a given rule at a given position. How do I do that?

**Comment by sogaiu** [comment, marked as answer]
I don't know about figuring out the precise rule, but the `--debug` and `--debug-graph` (this one makes a `log.html` file with svg in it) flags of the `parse` subcommand for the `tree-sitter` cli gives some detailed info about parsing.

I have some introductory notes on these bits [here](https://github.com/sogaiu/tree-sitter-clojure/blob/dev-notes/doc/dev-notes.md#tree-sitter-parse---debug) and [here](https://github.com/sogaiu/tree-sitter-clojure/blob/dev-notes/doc/dev-notes.md#tree-sitter-parse---debug-graph).  Not sure how much help those will be but FWIW.

There's also a `--report-states-for-rule` option for the `generate` subcommand, but I'm not sure what it does in detail.
